File: 剑指offer/06-从尾到头打印链表P58/Solution.cpp
// Author : lihongyu
// Time   : 2019-10-28

#include<bits/stdc++.h>
using namespace std;

struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

ListNode* CreateList(int a[], int n){
	
	if(n == 0)	return NULL;
	
	ListNode* head = new ListNode(a[0]);
	ListNode* curNode = head;
	
	for(int i = 1; i < n; i++){
		curNode->next = new ListNode(a[i]);
		curNode = curNode->next;
	}
	return head;
}

void PrintList(ListNode* head){
	
	ListNode* curNode = head;
	while(curNode != NULL){
		cout<<curNode->val<<" -> ";
		curNode = curNode->next;
	}
	cout<<"NULL"<<endl;
	return;
}

void DeleteList(ListNode* head){
	
	ListNode* curNode = head;
	while(curNode != NULL){
		ListNode* delNode = curNode;
		curNode = curNode->next;
		delete delNode;
	}
	
	return;
}
//reverse list via stack
void IterativePrintListReverse(ListNode* head){
	
	stack<ListNode*> nodes;
	
	ListNode* curNode = head;
	while(curNode != NULL){
		
		nodes.push(curNode);
		curNode = curNode->next;
	}
	while(!nodes.empty()){
		curNode = nodes.top();
		cout<<curNode->val<<" -> ";
		nodes.pop();
	}
	cout<<"NULL"<<endl;
	return;
}

void RecursivePrintListReverse(ListNode* head){
	
	if(head != NULL){
		if(head->next != NULL){
			RecursivePrintListReverse(head->next);
		}
		cout<<head->val<<" -> ";
	}
	return;
} 

//leetcode-206
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* pre = NULL;
		ListNode* cur = head;
		while(cur != NULL){
			ListNode* next = cur->next;
			
			cur->next = pre;
			pre = cur;
			cur = next;
		} 
		return pre;
    }
};

int main(){
	
	int a[] = {1, 2, 3, 4, 5};
	
	ListNode* head = CreateList(a, 5);
	PrintList(head);
	
//	ListNode* h2 = Solution().reverseList(head);
//	PrintList(h2);
	
//	IterativePrintListReverse(head);

	RecursivePrintListReverse(head);
	cout<<"NULL"<<endl;
	
//	DeleteList(h2);
	return 0;
} 

File: 剑指offer/06-从尾到头打印链表P58/Solution.java
// Author : lihongyu
// Time   : 2020-2-14
// leetcode-面试题06. 从尾到头打印链表
// 思路：链表
// 时间复杂度: O(n)
// 空间复杂度: O(1)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

class Solution {
    public ListNode CreateList(int[] a){
        int n = a.length;
        if(n == 0)	return null;

        ListNode head = new ListNode(a[0]);
        ListNode curNode = head;

        for(int i = 1; i < n; i++){
            curNode.next = new ListNode(a[i]);
            curNode = curNode.next;
        }
        return head;
    }

    public int[] reversePrint(ListNode head) {

        int i = 0, j = 0;
        ListNode pre = null;
		ListNode cur = head;
		while(cur != null){
			ListNode next = cur.next;
			cur.next = pre;
			pre = cur;
			cur = next;
			i++;
		}
		int[] res = new int[i];
		while(pre != null){
            res[j++] = pre.val;
            pre = pre.next;
        }
		return res;
    }

    public static void printArray(int[] nums){
        for(Integer e: nums)
            System.out.print(e + " ");
        System.out.println();
    }

    public static void main(String[] args) {

        int[] a = {1, 3, 2};
        ListNode head = (new Solution()).CreateList(a);
        int[] res = (new Solution()).reversePrint(head);
        printArray(res);

    }
}


// TODO: